kuma/users/signal_handlers.py

```diff
@@ -1,23 +1,73 @@
-from allauth.account.signals import email_confirmed, user_signed_up
+from allauth.account.signals import (
+    email_confirmed,
+    user_logged_in,
+    user_signed_up)
+from allauth.socialaccount.signals import pre_social_login
 from allauth.socialaccount.signals import social_account_removed
 from django.core.exceptions import ObjectDoesNotExist
 from django.db import transaction
 from django.db.models.signals import post_delete, post_save, pre_delete
 from django.dispatch import receiver
 from waffle import switch_is_active
 
+from kuma.core.ga_tracking import (
+    ACTION_AUTH_SUCCESSFUL,
+    ACTION_FREE_NEWSLETTER,
+    ACTION_PROFILE_CREATED,
+    ACTION_RETURNING_USER_SIGNIN,
+    CATEGORY_SIGNUP_FLOW,
+    track_event)
 from kuma.payments.utils import cancel_stripe_customer_subscription
 from kuma.wiki.jobs import DocumentContributorsJob
 
 from .models import User, UserBan
 from .tasks import send_welcome_email
 
 
+@receiver(pre_social_login, dispatch_uid='users.pre_social_login')
+def on_pre_social_login(sender, request, **kwargs):
+    """
+    Signal handler to be called when a given user has at least successfully
+    authenticated with a provider but not necessarily fully logged in on
+    our site. For example, if the user hasn't created a profile (e.g.
+    agreeing to terms and conditions) the 'user_logged_in' won't fire until
+    then.
+    """
+    # If the user did the "social_auth_add" they already logged in and
+    # all we needed to do was to "combine" their user (from two different
+    # providers) can be reduce to the same.
+    if getattr(request, 'social_auth_added', False):
+        return
```

Comment:
This is the right thing in the wrong place. It should be moved to the beginning of the `on_user_signed_up()` handler immediately below.